Simplifies API of WireCoders and LengthPrefixUnknownCoders.

jkff opened the pull request:
Instead of returning MessageWithComponents, they take a Components.Builder directly and add the necessary coders to it. This considerably simplifies both the implementation code and caller code: see ProcessBundleDescriptors.

I had to refactor ProcessBundleDescriptors a bit to make this work: I'm not sure why ProcessBundleDescriptor doesn't just have a Components inside it already, and instead has the same fields - but using Components at least for internal calculations simplifies that code too.

R: @lukecwik (sorry for bombarding you with reviews today: FWIW I'm not blocked on this PR, it can wait a few days if you're busy)

jkff commented:
CC: @bsidhom

jkff commented:
Run Java PreCommit

jkff commented:
retest this please

lukecwik commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlers.java:
```diff
@@ -178,10 +179,10 @@ static BagUserStateHandlerFactory unsupported() {
    * {@link MultimapSideInputHandlerFactory} are cached.
    */
   public static StateRequestHandler forMultimapSideInputHandlerFactory(
-      ExecutableProcessBundleDescriptor processBundleDescriptor,
```
Why did you want to break this chain by not passing in the ExecutableProcessBundleDescriptor?

Also, did you meant to make this change within this PR?

bsidhom replied:
ExecutableProcessBundleDescriptor contains API descriptors for the state and data services. We do not have access to those in the places where we construct state handlers in the Flink runner and in general those should not be required when defining handler behavior.

lukecwik commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/wire/LengthPrefixUnknownCoders.java:
```diff
@@ -16,43 +16,36 @@
+/** Utilities for replacing or wrapping unknown coders with {@link LengthPrefixCoder}. */
 public class LengthPrefixUnknownCoders {
   /**
    * Recursively traverse the coder tree and wrap the first unknown coder in every branch with a
```
Is there a way to make this comment clearer that the components builder is modified?

lukecwik commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/wire/LengthPrefixUnknownCoders.java:
```diff
@@ -62,109 +55,69 @@
-        coderId + "-byte_array",
-        Sets.union(components.getCodersMap().keySet(),
-            rvalBuilder.getComponents().getCodersMap().keySet()));
+  /** Adds the (singleton) length-prefixed byte array coder. */
```
This will get added multiple times so its not really a singleton

lukecwik replied:
Please fix the comment since I still think this applies.

jkff replied:
Not anymore - I changed the code to not add coders that are already in the map.

lukecwik commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/wire/WireCoders.java:
```diff
@@ -38,11 +36,11 @@
    * unknown to the runner are wrapped with length-prefix coders. The inner element coders are kept
```
Update comment to state that the components builder is modified to contain the new coder.

lukecwik commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/wire/WireCoders.java:
```diff
@@ -51,9 +49,9 @@
    *
```
Fix comment like above to say we modify the components builder and also update what we return

lukecwik replied:
Please fix the comment since I still think this applies.

jkff replied:
Not sure what you mean: I added a paragraph to the comment saying this adds coders to the builder, and modified `@return` to say it returns the id of the new coder.

lukecwik commented:
This might break Dataflow as the API changed might be shared with the internal implementation.

jkff commented:
Addressed all comments, thanks, PTAL. Regarding Dataflow: sure, next time when building the Dataflow worker someone will need to make the simple changes to the Dataflow runner harness - but this shouldn't break any existing jobs because this is not SDK code.

jkff commented:
I also updated the Flink batch side input handler - it is no longer awkwardly constructing an executable ProcessBundleDescriptor.
